src/Catalog/Icons/CatalogLeafDataProcessor.cs

```diff
@@ -0,0 +1,274 @@
+            _iconProcessor = iconProcessor ?? throw new ArgumentNullException(nameof(iconProcessor));
+            _externalIconContentProvider = externalIconContentProvider ?? throw new ArgumentNullException(nameof(externalIconContentProvider));
+            _iconCopyResultCache = iconCopyResultCache ?? throw new ArgumentNullException(nameof(iconCopyResultCache));
+            _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task ProcessPackageDeleteLeafAsync(Storage storage, CatalogCommitItem item, CancellationToken cancellationToken)
+        {
+            var targetStoragePath = GetTargetStorageIconPath(item);
+            await _iconProcessor.DeleteIcon(storage, targetStoragePath, cancellationToken, item.PackageIdentity.Id, item.PackageIdentity.Version.ToNormalizedString());
+            // it would be nice to remove the icon copy result from cache for this item, but we don't have an icon URL here,
+            // so can't remove anything. Will rely on the copy code to catch the copy failure and cleanup the cache appropriately.
+        }
+
+        public async Task ProcessPackageDetailsLeafAsync(IStorage destinationStorage, CatalogCommitItem item, string iconUrlString, string iconFile, CancellationToken cancellationToken)
+        {
+            var hasExternalIconUrl = !string.IsNullOrWhiteSpace(iconUrlString);
+            var hasEmbeddedIcon = !string.IsNullOrWhiteSpace(iconFile);
+            if (hasExternalIconUrl && !hasEmbeddedIcon && Uri.TryCreate(iconUrlString, UriKind.Absolute, out var iconUrl))
+            {
+                using (_logger.BeginScope("Processing icon url {IconUrl}", iconUrl))
+                {
+                    await ProcessExternalIconUrlAsync(destinationStorage, item, iconUrl, cancellationToken);
+                }
+            }
+            else if (hasEmbeddedIcon)
+            {
+                await ProcessEmbeddedIconAsync(destinationStorage, item, iconFile, cancellationToken);
+            }
+        }
+
+        private async Task ProcessExternalIconUrlAsync(IStorage destinationStorage, CatalogCommitItem item, Uri iconUrl, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Found external icon url {IconUrl} for {PackageId} {PackageVersion}",
+                iconUrl,
+                item.PackageIdentity.Id,
+                item.PackageIdentity.Version);
+            if (!IsValidIconUrl(iconUrl))
+            {
+                _logger.LogInformation("Invalid icon URL {IconUrl}", iconUrl);
+                return;
+            }
+            var cachedResult = _iconCopyResultCache.Get(iconUrl);
+            if (cachedResult != null)
+            {
+                if (cachedResult.IsCopySucceeded)
+                {
```

Comment:
This method is quite complex. Consider splitting copying from cached result and copying externally:

```
if (cachedResult != null && await TryCopyFromCachedResultAsync(cachedResult))
{
   return;
}

// Copy externally..
```